platform/posix/transport/src/openssl_posix.c

```diff
@@ -0,0 +1,563 @@
+            sniHostName[ pOpensslCredentials->sniHostNameLen ] = '\0';
+            ( void ) memcpy( sniHostName,
+                             pOpensslCredentials->sniHostName,
+                             pOpensslCredentials->sniHostNameLen );
+
+            sslStatus = SSL_set_tlsext_host_name( pNetworkContext->pSslContext,
+                                                  sniHostName );
+
+            if( sslStatus != 1 )
+            {
+                LogError( ( "Failed to set server name %.*s for SNI.",
+                            ( int32_t ) pOpensslCredentials->sniHostNameLen,
+                            pOpensslCredentials->sniHostName ) );
+            }
+        }
+    }
+
+    /* Perform the TLS handshake. */
+    if( returnStatus == OPENSSL_SUCCESS )
+    {
+        sslStatus = SSL_connect( pNetworkContext->pSslContext );
+
+        if( sslStatus != 1 )
+        {
+            LogError( ( "SSL_connect failed to perform TLS handshake." ) );
+            returnStatus = OPENSSL_HANDSHAKE_FAILED;
+        }
+    }
+
+    /* Verify X509 certificate from peer. */
+    if( returnStatus == OPENSSL_SUCCESS )
+    {
+        verifyPeerCertStatus = SSL_get_verify_result( pNetworkContext->pSslContext );
+
+        if( verifyPeerCertStatus != X509_V_OK )
+        {
+            LogError( ( "SSL_get_verify_result failed to verify X509 "
+                        "certificate from peer." ) );
+            returnStatus = OPENSSL_HANDSHAKE_FAILED;
+        }
+    }
+
+    /* Clean up on error. */
+    if( ( returnStatus != OPENSSL_SUCCESS ) && ( sslStatus == 0 ) )
+    {
+        SSL_free( pNetworkContext->pSslContext );
+        pNetworkContext->pSslContext = NULL;
+    }
+
+    /* Log failure or success depending on status. */
+    if( returnStatus != OPENSSL_SUCCESS )
+    {
+        LogError( ( "Failed to establish a TLS connection." ) );
+    }
+    else
+    {
+        LogDebug( ( "Established a TLS connection." ) );
+    }
+
+    return returnStatus;
+}
+
+OpensslStatus_t Openssl_Disconnect( NetworkContext_t pNetworkContext )
+{
+    OpensslStatus_t returnStatus = OPENSSL_SUCCESS;
+
+    if( pNetworkContext == NULL )
+    {
+        LogError( ( "Parameter check failed: pNetworkContext is NULL." ) );
+        returnStatus = OPENSSL_INVALID_PARAMETER;
+    }
+    else if( pNetworkContext->pSslContext != NULL )
+    {
+        /* SSL shutdown should be called twice: once to send "close notify" and
+         * once more to receive the peer's "close notify". */
+        if( SSL_shutdown( pNetworkContext->pSslContext ) == 0 )
+        {
+            ( void ) SSL_shutdown( pNetworkContext->pSslContext );
+        }
+
+        SSL_free( pNetworkContext->pSslContext );
+    }
+    else
+    {
+        /* Empty else for MISRA 15.7 compliance. */
+    }
+
+    return returnStatus;
+}
+
+void Openssl_SetRecvTimeout( int timeout )
+{
+    tlsRecvTimeout = timeout;
+}
+
+void Openssl_SetSendTimeout( int timeout )
```

Comment:
Where is this function and above function is used ? Are they used by  the openSsl library or by the application ?

Reply:
As discussed offline, we are removing the ability to set timeouts with a call to a function. User will only be able to set timeout when calling `TCP_Connect`.